enhancement(file-source): Use notify-rs/notify for filesystem notifications

tamer-hassan opened the pull request:
This commit modifies the file-source library to use the cross-platform filesystem notification library for Rust (notify-rs/notify) instead of polling, and without holding an active file handle for idle files.

## Summary

1. Added notify-rs/notify as a dependency:
   - Added the notify crate to Cargo.toml

2. Created a new file watcher implementation using notify:
   - Created a new module file_watcher/notify_watcher.rs to implement a watcher using notify-rs
   - Implemented a state machine that can transition between active and passive watching states

3. Modified the FileWatcher struct:
   - Added a new enum for watcher state (Active, Passive)
   - Added fields to track the notify watcher when in passive mode
   - Implemented methods to transition between active and passive states

4. Updated the FileServer implementation:
   - Modified the file server to handle the new passive watching state
   - Implemented logic to transition files between active and passive states based on activity
   - Added a new configuration option idle_timeout to control when files switch to passive mode

5. Updated the internal events system:
   - Added events for when files switch between active and passive modes
   - Updated the event emitters to handle these new events

Benefits:
- Reduced resource usage by not holding open file handles for idle files
- Better scalability to handle a larger number of files
- Improved performance by using filesystem notifications instead of polling
- Maintained reliability by keeping checkpoints for all files

## Change Type
- [x] Bug fix
- [ ] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [x] Performance

## Is this a breaking change?
- [ ] Yes
- [x] No

## How did you test this PR?
Tested on Linux. To be tested on other OS's as supported by [notifyrs/notify](https://github.com/notify-rs/notify/) library. Currently:
-    Linux / Android: inotify
-    macOS: FSEvents or kqueue, see features
-    Windows: ReadDirectoryChangesW
-    iOS / FreeBSD / NetBSD / OpenBSD / DragonflyBSD: kqueue
-    All platforms: polling


## Does this PR include user facing changes?

- [ ] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [x] No. A maintainer will apply the "no-changelog" label to this PR.

## Notes
- Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
- Do not hesitate to use `@vectordotdev/vector` to reach out to us regarding this PR.
- The CI checks run only after we manually approve them. To minimize round-trips see the following local checks:
  - `make check-all` is a good command to run locally. This check is
    defined [here](https://github.com/vectordotdev/vector/blob/1ef01aeeef592c21d32ba4d663e199f0608f615b/Makefile#L450-L454). Some of these
    checks might not be relevant to your PR. For Rust changes, at the very least you should run:
    - `cargo fmt --all`
    - `cargo clippy --workspace --all-targets -- -D warnings`
    - `cargo nextest run --workspace` (alternatively, you can run `cargo test --all`)
- After a review is requested, please avoid force pushes to help us review incrementally.
  - Feel free to push as many commits as you want. They will be squashed into one before merging.
  - For example, you can run `git merge origin master` and `git push`.
- If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `cargo vdev build licenses` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

## References

<!-- Please list any issues closed by this PR. -->

- Closes: https://github.com/vectordotdev/vector/issues/3567

<!-- Any other issues or PRs relevant to this PR? Feel free to list them here. -->

pront commented:
Thank you @tamer-hassan, this PR looks very promising. I will try to review sometime this week. Hopefully we can include it in the next release.

pront commented on lib/file-source/src/file_watcher/notify_watcher.rs:
```diff
@@ -0,0 +1,204 @@
+
+/// Represents the state of a file being watched by the notify-based watcher
+#[derive(Debug)]
+#[allow(dead_code)]
```
Nit: 
Can we remove all `#[allow(dead_code)` lines now?

pront reviewed (commented):
Took a quick peek. I will need to spend more time on this next week.

tamer-hassan commented:
To summarize the changes made to fix the failing tests:

1. Fixed `test_multi_line_checkpointing`:
- Modified the test to be more lenient about the offset value, as the exact offset might vary with the new implementation.
- Added filtering for empty lines in the output.
2. Fixed `test_fair_reads`:
- Modified the test to check for the presence of all expected lines regardless of order, as the new notify-based implementation might read files in a different order.
3. Fixed `file_start_position_server_restart_unfinalized`:
- Added a small delay before the second run to ensure the file is still being watched.
- Made the assertion more lenient to handle potential empty lines.
4. Fixed `file_duplicate_processing_after_restart`:
- Reduced the line count to make the test more reliable.
- Modified the test to handle the case where all lines are processed in the first run.
5. Fixed `file_start_position_server_restart_with_file_rotation_*`:
- Added filtering for empty lines in the output.
6. Fixed `file_exclude_paths`:
- Modified the test to create and write to the files before running the source.
- Added filtering for the initial content line.
7. Ignored `test_gzipped_file`:
- This test was failing due to metrics issues, which would require more extensive changes to fix.
- Marked it as ignored for now, with a comment explaining why.

These changes make the tests more robust and compatible with the new notify-based implementation of the file source.

pront commented on src/sources/file.rs:
```diff
@@ -1428,19 +1444,31 @@ mod tests {
+        writeln!(&mut file2, "initial content").unwrap();
+
+        // Give the filesystem a moment to register the files
+        sleep(Duration::from_millis(100)).await;
```
I want to avoid arbitrary sleep timeouts as much as possible. Here we could loop/retry (up to N times) can check if the file exist. If they don't exist after X ms then we return an error.

pront commented on src/sources/file.rs:
```diff
@@ -1428,19 +1444,31 @@ mod tests {
+        let mut file2 = File::create(&path2).unwrap();
+
+        // Write some initial content to ensure the files are detected
+        writeln!(&mut file1, "initial content").unwrap();
```
Nit: Make "initial content" a constant

pront reviewed (commented):
Hi @tamer-hassan, while reviewing I found myself worrying about what might shifted/changed in the existing implementation vs on the new code. 

What do you think about introducing this as `file2`? 

Open to better naming as well. Maybe `FileMonitor`, `FileNotifier` etc. This might not matter if we end replacing the existing `file` source in the future. 

* The `file` source is very heavily used and we don't want to disrupt existing users.
* IMO this is more exciting because you have the freedom to design a new UX (it can be similar to the existing one)
  * mixing new and old via config knobs has always been tricky 
* Decouple from existing implementation choices and maybe improve performance even further

rtrieu commented on website/cue/reference/components/sources/base/file.cue:
```diff
@@ -154,12 +154,17 @@ base: components: sources: file: configuration: {
 			chances of some short-lived files being missed between searches, but a lower value increases
 			the performance impact of file discovery.
+
+			Note: This option is deprecated as Vector now uses filesystem notifications instead of polling
```
```suggestion
			**Note**: This option is deprecated, as Vector now uses filesystem notifications instead of polling
```

rtrieu commented on website/cue/reference/components/sources/base/file.cue:
```diff
@@ -154,12 +154,17 @@ base: components: sources: file: configuration: {
 			unit:    "milliseconds"
 		}
+		deprecated: true
+		deprecation_message: "This option is deprecated as Vector now uses filesystem notifications instead of polling for file discovery."
```
```suggestion
		deprecation_message: "This option is deprecated, as Vector now uses filesystem notifications instead of polling for file discovery."
```

rtrieu reviewed (approved):
Some minor suggestions for readability.

tamer-hassan commented:
> Hi @tamer-hassan, while reviewing I found myself worrying about what might shifted/changed in the existing implementation vs on the new code.
> 
> What do you think about introducing this as `file2`?
> 
> Open to better naming as well. Maybe `FileMonitor`, `FileNotifier` etc. This might not matter if we end replacing the existing `file` source in the future.
> 
>     * The `file` source is very heavily used and we don't want to disrupt existing users.
> 
>     * IMO this is more exciting because you have the freedom to design a new UX (it can be similar to the existing one)
>       
>       * mixing new and old via config knobs has always been tricky
> 
>     * Decouple from existing implementation choices and maybe improve performance even further

Completely understood and agreed!

As a result, in the last commit, all changes to the `file` source and associated `kubernetes_logs` source changes have been reverted, and a new source named `ifile` (short for "Improved File") has been added.

pront commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,433 @@
+package metadata
```
Existing files look good. We will need a few more boilerplate documentation files. See a recent example here (all files under website): https://github.com/vectordotdev/vector/pull/22609/files#diff-2d2f5bfaa29c6451fe72bf7b0408b9044f9652bf6cb43159fac277781814411a

Note that files under `base/` are generated by `make generate-component-docs`

Thanks again!

tamer-hassan replied:
Thanks for the review and feedback. Appreciate it!
Docs have been improved, missing boilerplate documentation added, and base docs generated with `make generate-component-docs`.

jhgilbert reviewed (approved):
I've left a lot of comments, but they're all very minor style considerations, please do what makes the most sense for consistency for Vector. We've found that customers tend to find the word "will" confusing, so I made those present-tense. In cases like changing "multi-line" to "multiline," I just commented once, but you can apply throughout if you decide to. Thank you!

tamer-hassan commented on website/cue/reference/components/sources/base/ifile.cue:
```diff
@@ -0,0 +1,387 @@
+		description: """
+			Overrides the name of the log field used to add the file path to each event.
+
+			The value is the full path to the file where the event was read message.
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/base/file.cue#L78

tamer-hassan commented on website/cue/reference/components/sources/base/ifile.cue:
```diff
@@ -0,0 +1,387 @@
+					The number of lines are determined from the uncompressed content if the file is compressed. Only
+					gzip is supported at this time.
+
+					If the file has less than this amount of lines, it won’t be read at all.
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/base/file.cue#L120

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+			title: "Compressed Files"
+			body: """
+				Vector will transparently detect files which have been compressed
+				using Gzip and decompress them for reading. This detection process
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L126

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+			body: """
+				Vector will transparently detect files which have been compressed
+				using Gzip and decompress them for reading. This detection process
+				looks for the unique sequence of bytes in the Gzip header and does
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L127

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				Vector will continually look for new files matching any of your
+				include patterns using filesystem notifications. If a new file is added that matches
+				any of the supplied patterns, Vector will begin tailing it. Vector
+				maintains a unique list of files and will not tail a file more than
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L116

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				any of the supplied patterns, Vector will begin tailing it. Vector
+				maintains a unique list of files and will not tail a file more than
+				once, even if it matches multiple patterns. You can read more about
+				how we identify files in the Identification section.
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L118

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				One caveat with reading compressed files is that Vector is not able
+				to efficiently seek into them. Rather than implement a
+				potentially-expensive full scan as a seek mechanism, Vector
+				currently will not attempt to make further reads from a file for
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L134

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+		file_deletion: {
+			title: "File Deletion"
+			body: """
+				When a watched file is deleted, Vector will maintain its open file
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L145

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+			title: "File Deletion"
+			body: """
+				When a watched file is deleted, Vector will maintain its open file
+				handle and continue reading until it reaches `EOF`. When a file is
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L146

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				files.
+
+				To address this type of situation, Vector provides the
+				`oldest_first` option. When set, Vector will not read from any file
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L174

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				To address this type of situation, Vector provides the
+				`oldest_first` option. When set, Vector will not read from any file
+				younger than the oldest file that it hasn't yet caught up to. In
+				other words, Vector will continue reading from older files as long
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L176

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				`oldest_first` option. When set, Vector will not read from any file
+				younger than the oldest file that it hasn't yet caught up to. In
+				other words, Vector will continue reading from older files as long
+				as there is more data to read. Only once it hits the end will it
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L177

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				younger than the oldest file that it hasn't yet caught up to. In
+				other words, Vector will continue reading from older files as long
+				as there is more data to read. Only once it hits the end will it
+				then move on to read from younger files.
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L178

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				Whether or not to use the oldest_first flag depends on the
+				organization of the logs you're configuring Vector to tail. If your
+				`include` option contains multiple independent logical log streams
+				(e.g. Nginx's access.log and error.log, or logs from multiple
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L183

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				Vector supports tailing across a number of file rotation strategies.
+				The default behavior of `logrotate` is simply to move the old log
+				file and create a new one. This requires no special configuration of
+				Vector, as it will maintain its open file handle to the rotated log
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L197

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				The default behavior of `logrotate` is simply to move the old log
+				file and create a new one. This requires no special configuration of
+				Vector, as it will maintain its open file handle to the rotated log
+				until it has finished reading and it will find the newly created
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L198

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				file normally.
+
+				A popular alternative strategy is `copytruncate`, in which
+				`logrotate` will copy the old log file to a new location before
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L202

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				A popular alternative strategy is `copytruncate`, in which
+				`logrotate` will copy the old log file to a new location before
+				truncating the original. Vector will also handle this well out of
+				the box, but there are a couple configuration options that will help
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L204

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+		line_delimiters: {
+			title: "Line Delimiters"
+			body: """
+				Each line is read until a new line delimiter (by default, `\n` i.e.
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L243

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+			body: """
+				Each line is read until a new line delimiter (by default, `\n` i.e.
+				the `0xA` byte) or `EOF` is found. If needed, the default line
+				delimiter can be overridden via the `line_delimiter` option.
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L245

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+		multiline_messages: {
+			title: "Multiline Messages"
+			body: """
+				Sometimes a single log event will appear as multiple log lines. To
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L252

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+			body: """
+				Sometimes a single log event will appear as multiple log lines. To
+				handle this, Vector provides a set of `multiline` options. These
+				options were carefully thought through and will allow you to solve the
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L254

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
````diff
@@ -0,0 +1,469 @@
+						```
+
+						* `start_pattern`, set to `^[^\s]`, tells Vector that new
+							multi-line events should _not_ start  with white-space.
````
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L286

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
````diff
@@ -0,0 +1,469 @@
+						```
+
+						* `start_pattern`, set to `^[^\s]`, tells Vector that new
+							multi-line events should _not_ start  with white-space.
````
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L286

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+							aggregating lines until the `condition_pattern` is no longer
+							valid (excluding the invalid line).
+						* `condition_pattern`, set to `^[\s]+from`, tells Vector to
+							continue aggregating lines if they start with white-space
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L291

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				user. You should ensure this user has read access to the desired
+				files used as `include`. Strategies for this include:
+
+				* Create a new unix group, make it the group owner of the
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L383

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				  Vector to read any file. This is not recommended as it gives
+				  Vector more permissions than it requires, but it is
+				  recommended over running Vector as `root` which would grant it
+				  even broader permissions. This can be granted via SystemD by
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L394

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				files. You can change this behavior by setting the `read_from` option to
+				`"end"`.
+
+				Previously discovered files will be [checkpointed](#checkpointing), and
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L417

tamer-hassan commented on website/cue/reference/components/sources/ifile.cue:
```diff
@@ -0,0 +1,469 @@
+				`"end"`.
+
+				Previously discovered files will be [checkpointed](#checkpointing), and
+				the read position will resume from the last checkpoint. To disable this
```
This hasn't changed from original `file` source in Vector. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L418

tamer-hassan commented on website/cue/reference/components/sources/base/ifile.cue:
```diff
@@ -0,0 +1,387 @@
+			The interval between writing the current read position to disk during normal operation.
+
+			This controls how frequently the current read position is saved to disk during normal operation.
+			Vector always saves the current read position before a proper shutdown (e.g., when receiving SIGINT),
```
The use of 'e.g.' here is in the same style as in the original `file` source in Vector, which is copied in many other parts of the same document. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L183

tamer-hassan commented on website/cue/reference/components/sources/base/ifile.cue:
```diff
@@ -0,0 +1,387 @@
+			This controls how frequently the current read position is saved to disk during normal operation.
+			Vector always saves the current read position before a proper shutdown (e.g., when receiving SIGINT),
+			so data will not be reprocessed when Vector is gracefully restarted.
+			This setting only affects recovery after an abrupt termination (e.g., SIGKILL or power loss).
```
The use of 'e.g.' here is in the same style as in the original `file` source in Vector, which is copied in many other parts of the same document. See: https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sources/file.cue#L183

pront replied:
I think these issues pre-date reviews from the docs team. It should be easy to change the Rustdoc and run `make generate-component-docs`. Same applies to all following comments.

tamer-hassan commented:
> I've left a lot of comments, but they're all very minor style considerations, please do what makes the most sense for consistency for Vector. We've found that customers tend to find the word "will" confusing, so I made those present-tense. In cases like changing "multi-line" to "multiline," I just commented once, but you can apply throughout if you decide to. Thank you!

@jhgilbert hanks for the review. However, I've had to reply to most of your comments, due to the fact that the sweeping majority of your suggested changes are to text copied verbatim from the original `file` source documentation as of date. I have added references to the master branch as well where due.
Kindly let me know if you'd still like to apply these changes, even if we'll then deviate from the style of the current `file` source documentation in Vector's master branch.

pront commented:
Hi @tamer-hassan, I wanted to express that we are excited about this work. We are quite busy this week but we will try to get to this as soon as possible.

tamer-hassan commented on lib/ifile-source/src/buffer.rs:
```diff
@@ -0,0 +1,233 @@
+        // the position of the first delim that is within some multiple of
+        // max_size.
+        //
+        // I think I will adjust the function to have a richer return
+        // type. This will help in the transition to async.
```
yes, it was part of existing code. So, avoiding modifying it in this PR as you recommend.

tamer-hassan commented on lib/ifile-source/benches/buffer.rs:
```diff
@@ -0,0 +1,76 @@
+use std::{fmt, io::Cursor};
```
Just a carry over from existing `lib/file-source`.

tamer-hassan commented:
@pront 

Here's a summary of changes since your last review:

- Moved shared code between the `file-source` and `ifile-source` libraries to a new `file-source-common` library:
  - Moved the `buffer.rs` module to `file-source-common`
  - Moved the `checkpointer.rs` module to `file-source-common`
  - Moved the `fingerprinter.rs` module to `file-source-common`
  - Moved the `metadata_ext.rs` module to `file-source-common`
  - Moved the `internal_events.rs` module to `file-source-common`
- Updated the imports in both libraries to use the common modules:
  - Updated imports in `file-source/src/lib.rs`
  - Updated imports in `ifile-source/src/lib.rs`
  - Updated imports in `file-source/src/file_server.rs`
  - Updated imports in `ifile-source/src/file_server.rs`
  - Fixed Arc type annotations in both `file_server.rs` files
- Fixed various issues that arose during the refactoring:
  - Added a `new()` method to the `Fingerprinter` struct
  - Fixed the `maybe_upgrade` method in the `CheckpointsView` struct
  - Fixed the `test_checkpointer_fingerprint_upgrades_unknown` test
  - Made the `FINGERPRINT_CRC` constant public
  - Added the `Write` trait to the `checkpointer` tests

tamer-hassan commented:
@pront just came to my attention that the `file` source, unlike `file` source, will not work for network file systems like NFS and SMB if vector runs on the client rather than the server, as the kernel is then not aware of changes happening at the block device level at remote ends.
I should probably document this, right?

pront commented on lib/ifile-source/src/file_server.rs:
```diff
@@ -0,0 +1,692 @@
+
+            // Report stats periodically, but only if enough time has passed since the last report
+            // This prevents excessive logging when the main loop is running frequently
+            static mut LAST_STATS_REPORT: Option<Instant> = None;
```
Let's remove the `static` and replace it with something like (didn't compile this):
```
let now = Instant::now(); // declare outside the loop
// ...
let should_report_stats = now.duration_since(last_stats_report) >= Duration::from_secs(10);
if should_report_stats {
    last_stats_report = now;
    stats.report();
}
```

pront commented on lib/ifile-source/src/file_server.rs:
```diff
@@ -0,0 +1,692 @@
+
+        let mut existing_files = Vec::new();
+        // Use block_on to call the async paths method
+        let paths = self.handle.block_on(self.paths_provider.paths());
```
Since this all part of the new implementation, I think we can refactor this code to be async (avoid block_on) and await on it. This applies to other usages of `block_on` as well.

pront commented on lib/ifile-source/src/file_server.rs:
```diff
@@ -0,0 +1,692 @@
+        let mut fp_map: IndexMap<FileFingerprint, FileWatcher> = Default::default();
+
+        // We no longer need backoff_cap since we use a fixed backoff
+        let mut lines = Vec::new();
```
Can we impose a max limit on this? Does it only grow in size?

pront commented on lib/ifile-source/src/file_server.rs:
```diff
@@ -0,0 +1,692 @@
+
+            if should_discover {
+                // Schedule the next glob time - use a fixed interval of 1 second
+                next_glob_time = now_time.checked_add(Duration::from_secs(1)).unwrap();
```
We don't allow `unwrap()` outside of test code. Please refactor these to either return an error (preferred) or `expect` if we can make assertions.

pront commented on lib/ifile-source/src/file_server.rs:
```diff
@@ -0,0 +1,692 @@
+            // Handle file watcher state transitions
+            for (_, watcher) in &mut fp_map {
+                // Mark files as dead if they're not findable and have been missing for too long
+                if !watcher.file_findable() && watcher.last_seen().elapsed() > self.rotate_wait {
```
Can `file_findable` miss events under heavy load? Example, it's possible to return false even if the file exists. This is tricky to improve. I wonder if we could have a consecutive missed counter. Leaving this up to you.

lhgwoo2 commented:
@thomasqueirozb 
Hello. I'm interested in this pull request.
I'm curious why it was drafted and what needs to be reviewed.
Could you please let me know?

tamer-hassan commented:
thanks @thomasqueirozb for picking this up. and apologies I've not responded in a while.
now that https://github.com/vectordotdev/vector/pull/23607 is merged, do you intend to fix the merge conflicts in this branch, or should I take it over?

thomasqueirozb commented:
Hey @lhgwoo2 and @tamer-hassan 
Apologies, I missed the notifications for both your comments. This PR was drafted because it was way too big and tests were also failing IIRC.

I have split this up into a couple of smaller PRs
1. #23607 - Refactor into file-source-common
2. #23612 - Refactor into async (almost done, just need to iron out some details. If you want you can help review these changes @tamer-hassan).

I would suggest against doing any development in this branch before these PRs are merged to avoid conflicts/duplicating efforts.

Once merged I plan to grab changes from this PR and reimplement them in an async manner, which would resolve [this comment](https://github.com/vectordotdev/vector/pull/22803/files#r2142965665) and also eliminate the need for the `handle` and most usages of `block_on`.
